### PR TITLE
회원가입 및 로그인

### PR DESCRIPTION
**Pull request by EunjungOh123**
**작업 내용**

☐ 로그인 기능 구현
☐ 회원 가입 (유효성 검사 o, 이메일 인증 x) 구현

### CODE REVIEW

**Comment by EunjungOh123** [comment]
|File|Coverage [0%]|:x:|
|:-|:-:|:-:|
|[RegisterMemberInput.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fdto%2FRegisterMemberInput.java)|0%|:x:|
|[Member.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fentity%2FMember.java)|0%|:x:|
|[BakingInProjectApplication.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2FBakingInProjectApplication.java)|0%|:x:|
|[MemberController.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fcontroller%2FMemberController.java)|0%|:x:|
|[CheckUserIdValidator.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fvalidator%2FRegisterValidatorInher%2FCheckUserIdValidator.java)|0%|:x:|
|[CheckEmailValidator.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fvalidator%2FRegisterValidatorInher%2FCheckEmailValidator.java)|0%|:x:|
|[CheckPasswordMatchValidator.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fvalidator%2FRegisterValidatorInher%2FCheckPasswordMatchValidator.java)|0%|:x:|
|[UserDetailsServiceImpl.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fservice%2FUserDetailsServiceImpl.java)|0%|:x:|
|[MemberServiceImpl.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fservice%2FMemberServiceImpl.java)|0%|:x:|
|[BaseTimeEntity.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fentity%2FBaseTimeEntity.java)|0%|:x:|
|[ServletInitializer.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fconfig%2FServletInitializer.java)|0%|:x:|
|[AuthFailureHandler.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fconfig%2FAuthFailureHandler.java)|0%|:x:|
|[SecurityConfig.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fconfig%2FSecurityConfig.java)|0%|:x:|
|[ErrorCode.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Ftype%2FErrorCode.java)|0%|:x:|
|[MemberRole.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Ftype%2FMemberRole.java)|0%|:x:|
|[MemberStatus.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Ftype%2FMemberStatus.java)|0%|:x:|
|[PasswordUtils.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Futil%2FPasswordUtils.java)|0%|:x:|
|[RegisterValidator.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fvalidator%2FRegisterValidator.java)|0%|:x:|
|[MemberStatusException.java](https://github.com/EunjungOh123/BakingIn-Project/blob/1a7d80968d5482b6b241b752b868b575a96650a9/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Fbakingin_project%2Fmember%2Fexception%2FMemberStatusException.java)|0%|:x:|

|Total Project Coverage|0%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/zerobase/bakingin_project/config/AuthFailureHandler.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+@Component
+public class AuthFailureHandler extends SimpleUrlAuthenticationFailureHandler {
+
+    @Override
+    public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response, AuthenticationException exception) throws IOException, ServletException {
+        String errorMessage;
+
+
+        if (exception instanceof UsernameNotFoundException) {
+            errorMessage = "존재하지 않는 계정입니다. 회원가입 후 로그인해주세요.";
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/config/AuthFailureHandler.java:24]: 이런 에러메시지 같은 경우는 enum 클래스나 추상클래스로 관리하시는 것도 나쁘지 않을 것 같다고 말씀드리려했는데, 아래 부분에 해당 클래스가 있었네요! 단지 이 부분은 초반에 만드시고 아직 수정을 안 하신거 같군요ㅎㅎ

File: src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+public class MemberController {
+
+    private final MemberService memberService;
+    private final CheckEmailValidator checkEmailValidator;
+    private final CheckUserIdValidator checkUserIdValidator;
+    private final CheckPasswordMatchValidator checkPasswordMatchValidator;
+
+    /* 중복 체크를 위한 커스텀 유효성 검사 */
+    /* 컨트롤러가 요청 올 때마다 WebDataBinder 호출되면서 WebDataBinder 에 등록한 검증기가 매번 적용 */
+
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java:30]: Validation을 이런식으로 처리하는 방법도 있군요!!

File: src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @PostMapping("/register")
+    public String registerSubmit (@Valid RegisterMemberInput memberInput, Errors errors, Model model) {
+        if(errors.hasErrors()) {
+            /* 회원가입 실패시 입력 데이터 유지 */
+            model.addAttribute("memberInput", memberInput);
+            /* 유효성 검사를 통과하지 못한 필드와 message 모델에 매핑해서 뷰로 전달 */
+            Map<String, String> validateMap = memberService.validateHandling(errors);
+            /* map.keySet() -> 모든 key 값 가져온 뒤 반복문을 통해 키와 에러 메세지로 매핑 */
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java:50]: 저는 컨트롤러에서 validation을 수행했었는데, 서비스 계층에서 구현하셨길래 이 방법이 좀 더 옳다고 생각해서 찾아보니 컨트롤러 보단 서비스, 서비스 보단 모델 계층에서 수행하는게 가장 좋다고 하네요! 그래도 서비스 계층에서 하는 것도 나쁘지 않아보입니다. 하나 배워갑니다!

File: src/main/java/com/zerobase/bakingin_project/member/dto/RegisterMemberInput.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    @NotBlank(message = "이름은 필수 입력 값입니다.")
+    @Size(min = 2, max = 8, message = "이름을 2~8자 사이로 입력해주세요.")
+    private String userName;
+
+    @NotBlank(message = "이메일은 필수 입력 값입니다.")
+    @Email(message = "이메일 형식이 올바르지 않습니다.")
+    private String email;
+
+    @Pattern(regexp = "(01[016789])(\\d{3,4})(\\d{4})", message = "- 없이 휴대폰 번호를 입력해주세요.")
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/member/dto/RegisterMemberInput.java:39]: 여기서도 몰랐던 것 배워갑니다~!

File: src/main/java/com/zerobase/bakingin_project/member/entity/Member.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import javax.persistence.*;
+import java.time.LocalDateTime;
+
+@Builder
+@Data
+@Entity
+@Table
+@NoArgsConstructor
+@AllArgsConstructor
+@Accessors(chain=true) //  Chain 형태로 이어서 원하는 set 메서드를 생성
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/member/entity/Member.java:21]: 와 이 방법은 정말 편한대요?! 저도 아프올 setter 사용할 일이 있으면 자주 애용할 것 같아요!

File: src/main/java/com/zerobase/bakingin_project/member/service/MemberServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                .role(MemberRole.GUEST)
+                .build();
+
+        memberRepository.save(member);
+
+        return true;
+    }
+
+    @Override
+    public Map<String, String> validateHandling(Errors errors) {
```
Comment (lyhjohn) [inline src/main/java/com/zerobase/bakingin_project/member/service/MemberServiceImpl.java:45]: 서비스계층에서 validation 구현하신 부분 배워 갑니다!

**Review by lyhjohn — COMMENTED** [review]
코드 잘 보고 갑니다! 특히 validation 처리가 인상 깊었네용. 고생하셨습니다!

File: src/main/java/com/zerobase/bakingin_project/member/type/ErrorCode.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.zerobase.bakingin_project.member.type;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+public enum ErrorCode {
```
Comment (spring-kotlin-lover) [inline src/main/java/com/zerobase/bakingin_project/member/type/ErrorCode.java:8]: MemberErrorCode 라고 네이밍을 수정해보시는 것은 어떨까요? 목적자체가 Member에 한정된 에러코드로 보입니다.

File: src/main/java/com/zerobase/bakingin_project/member/type/MemberStatus.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    STOP("USER_STATUS_STOP"),
+
+    /**
+     * 현재 탈퇴된 회원
+     */
+
+    WITHDRAW("USER_STATUS_WITHDRAW");
+
+    private final String value;
```
Comment (spring-kotlin-lover) [inline src/main/java/com/zerobase/bakingin_project/member/type/MemberStatus.java:36]: 내부 필드 변수에 대해 꼭 필요할까요? value 라는 필드가 어떤 역할을 하게 되는 것인지 궁금합니다.

File: src/main/java/com/zerobase/bakingin_project/member/service/MemberServiceImpl.java (hunk 6 above)
Comment (spring-kotlin-lover) [reply src/main/java/com/zerobase/bakingin_project/member/service/MemberServiceImpl.java:45]: validateHandling 에 대한 역할은 MemberService 에 있을 이유가 없다고 생각됩니다.
MemberService 라는 객체의 역할이 무엇일까요?

Member를 만들고, 조회하고, 삭제하고, 갱신하는 역할로 본다면 값 검증에 대한 validation 메서드가 MemberService 에 잇는 것은 Member 라는 비즈니스에 대해 슈퍼클래스를 만드는 것과 동일하다고도 볼 수 있을 것 같아요.

그렇다면 MemberService 의 책임은 높아지고, 역할이 많아지게 되겠죠.

값 검증에 대한 Validation 에 대한 역할만 잘 해내는 별도의 Validator 를 구현하시는 것을 추천드립니다.

**Review by spring-kotlin-lover — COMMENTED** [review]
안녕하세요~ 저는 무지멘토라고 하구요. 작성해주신 부분에 대해 리뷰드릴게 있어서 남깁니다.
전체적으로 코드 작성하시는 것을 보면 기본적인 구현능력이나 스프링에 대한 이해도가 있으신 것 같아요.

다만, 객체지향적인 설계에 있어서는 부족한점이 보여 리뷰를 드립니다.
객체가 무엇인지에 대한 아래 게시글을 참고하면 좋겠습니다.
https://imspear.tistory.com/187

File: src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java (hunk 2 above)
Comment (younghoondoodoom) [reply src/main/java/com/zerobase/bakingin_project/member/controller/MemberController.java:30]: validation에 대한 처리가 인상적이어서 보던 와중에 한가지 의문점이 들어서 말씀드리고자 합니다.
validation이 단순히 input값에 대한 validation이 아니라 이메일이 이미 존재하는지, 아이디가 이미 존재하는지에 대한 확인도 controller 단에서 처리하시는 것 같습니다.
제 생각에는 이 모든게 controller에서 처리가 된다면 확장하기 어려운 구조가 될 것 같습니다. 만약 MemberService를 이용하는 다른 컨트롤러(예를 들면 RestController로 바뀌는 경우) 여전히 이메일, 아이디의 검증은 해당 컨트롤러에서 진행해야합니다. 같은 코드가 중복된다는 뜻입니다.
그리고 요구사항에 따른 데이터의 무결성을 책임지는 부분은 service 단이라고 생각합니다. service 단에서 그런 검증을 해야 확장에 용이하지 않을까 생각합니다.

그래서 controller 단에서는 단순하게 input 값에 대한 validation(이메일 형식, 비밀번호 확인 일치, 아이디 형식 검사 등)만 진행하고 db를 사용하는 확인은 service 단에서 확인하는게 좋지 않을까 생각이 듭니다! 개인적인 의견이니 참고 부탁드립니다~

**Review by younghoondoodoom — COMMENTED** [review]
고생 하셨습니다~ validation 관련해 제 생각을 한번 말씀 드려봤습니다!

**Pull request merged**